File: edgequake/sdks/python/examples/error_handling.py
#!/usr/bin/env python3
"""
Error Handling — EdgeQuake Python SDK

WHY: Demonstrates how to handle different error types from the EdgeQuake SDK,
including retries, rate limiting, and graceful degradation.

Requirements:
    - EdgeQuake server running on http://localhost:8080 (may be offline for some tests)

Usage:
    export EDGEQUAKE_API_KEY="your-api-key"
    python examples/error_handling.py
"""
import os
import time

from edgequake import EdgequakeClient
from edgequake.exceptions import (
    EdgeQuakeError,
    NetworkError,
    NotFoundError,
    RateLimitedError,
    TimeoutError,
    UnauthorizedError,
    ValidationError,
)


def main():
    client = EdgequakeClient(
        api_key=os.environ.get("EDGEQUAKE_API_KEY", "your-api-key"),
        base_url="http://localhost:8080",
    )

    # --- Pattern 1: Specific error handling ---
    print("=== Pattern 1: Specific error types ===")
    try:
        client.documents.get("non-existent-id")
    except NotFoundError:
        print("Document not found — could prompt user to upload")
    except UnauthorizedError:
        print("Invalid API key — redirect to login")
    except RateLimitedError:
        print("Rate limited — back off and retry")
    except Exception as error:
        print(f"Unexpected error: {error}")

    # --- Pattern 2: Retry with exponential backoff ---
    print("\n=== Pattern 2: Retry with backoff ===")

    def query_with_retry(query: str, max_retries: int = 3):
        for attempt in range(1, max_retries + 1):
            try:
                return client.query.execute(query=query)
            except RateLimitedError:
                if attempt < max_retries:
                    delay = pow(2, attempt)  # 2s, 4s, 8s
                    print(
                        f"Rate limited, retrying in {delay}s "
                        f"(attempt {attempt}/{max_retries})"
                    )
                    time.sleep(delay)
                    continue
                raise  # Max retries exceeded
            except NetworkError:
                if attempt < max_retries:
                    print(
                        f"Network error, retrying " f"(attempt {attempt}/{max_retries})"
                    )
                    time.sleep(1)
                    continue
                raise  # Non-retryable or max retries exceeded

    try:
        result = query_with_retry("What is EdgeQuake?")
        print(f"Query result: {result.get('answer')}")
    except Exception as error:
        print(f"All retries exhausted: {error}")

    # --- Pattern 3: Graceful degradation ---
    print("\n=== Pattern 3: Graceful degradation ===")
    try:
        health = client.health()
        print(f"Backend healthy: {health.get('status')}")
    except (NetworkError, TimeoutError):
        print("Backend unreachable — showing cached data")
    except Exception as error:
        print(f"Unexpected health check failure: {error}")

    # --- Pattern 4: Validation error details ---
    print("\n=== Pattern 4: Validation errors ===")
    try:
        client.documents.upload(content="")  # Empty content
    except ValidationError as error:
        print(f"Validation failed: {error}")
        if hasattr(error, "status"):
            print(f"Status: {error.status}")  # 400 or 422
        if hasattr(error, "code"):
            print(f"Code: {error.code}")
    except Exception as error:
        print(f"Unexpected error: {error}")

    # --- Pattern 5: Generic EdgeQuakeError catch-all ---
    print("\n=== Pattern 5: Catch-all ===")
    try:
        client.documents.get("some-id")
    except EdgeQuakeError as error:
        # All SDK errors extend EdgeQuakeError
        status = getattr(error, "status", "N/A")
        code = getattr(error, "code", "N/A")
        print(f"API error [{status}] {code}: {error}")
    except Exception as error:
        # Non-SDK error (e.g., TypeError from bad config)
        print(f"Unexpected error: {error}")


if __name__ == "__main__":
    main()

File: edgequake/sdks/typescript/examples/error_handling.ts
/**
 * Example: Error Handling Patterns
 *
 * Demonstrates how to handle different error types from the EdgeQuake SDK,
 * including retries, rate limiting, and graceful degradation.
 */

import {
  EdgeQuake,
  EdgeQuakeError,
  NetworkError,
  NotFoundError,
  RateLimitedError,
  TimeoutError,
  UnauthorizedError,
  ValidationError,
} from "@edgequake/sdk";

async function main() {
  const client = new EdgeQuake({
    baseUrl: "http://localhost:8080",
    apiKey: "your-api-key",
  });

  // --- Pattern 1: Specific error handling ---
  console.log("=== Pattern 1: Specific error types ===");
  try {
    await client.documents.get("non-existent-id");
  } catch (error) {
    if (error instanceof NotFoundError) {
      console.log("Document not found — could prompt user to upload");
    } else if (error instanceof UnauthorizedError) {
      console.log("Invalid API key — redirect to login");
    } else if (error instanceof RateLimitedError) {
      console.log("Rate limited — back off and retry");
    } else {
      console.log("Unexpected error:", error);
    }
  }

  // --- Pattern 2: Retry with exponential backoff ---
  console.log("\n=== Pattern 2: Retry with backoff ===");
  async function queryWithRetry(query: string, maxRetries = 3) {
    for (let attempt = 1; attempt <= maxRetries; attempt++) {
      try {
        return await client.query.execute({ query });
      } catch (error) {
        if (error instanceof RateLimitedError && attempt < maxRetries) {
          const delay = Math.pow(2, attempt) * 1000; // 2s, 4s, 8s
          console.log(
            `Rate limited, retrying in ${delay}ms (attempt ${attempt}/${maxRetries})`,
          );
          await new Promise((r) => setTimeout(r, delay));
          continue;
        }
        if (error instanceof NetworkError && attempt < maxRetries) {
          console.log(
            `Network error, retrying (attempt ${attempt}/${maxRetries})`,
          );
          await new Promise((r) => setTimeout(r, 1000));
          continue;
        }
        throw error; // Non-retryable or max retries exceeded
      }
    }
  }

  try {
    const result = await queryWithRetry("What is EdgeQuake?");
    console.log("Query result:", result?.answer);
  } catch (error) {
    console.log("All retries exhausted:", error);
  }

  // --- Pattern 3: Graceful degradation ---
  console.log("\n=== Pattern 3: Graceful degradation ===");
  try {
    const health = await client.health();
    console.log("Backend healthy:", health.status);
  } catch (error) {
    if (error instanceof NetworkError || error instanceof TimeoutError) {
      console.log("Backend unreachable — showing cached data");
    } else {
      console.log("Unexpected health check failure:", error);
    }
  }

  // --- Pattern 4: Validation error details ---
  console.log("\n=== Pattern 4: Validation errors ===");
  try {
    await client.documents.upload({ content: "" }); // Empty content
  } catch (error) {
    if (error instanceof ValidationError) {
      console.log("Validation failed:", error.message);
      console.log("Status:", error.status); // 400 or 422
      console.log("Code:", error.code);
    }
  }

  // --- Pattern 5: Generic EdgeQuakeError catch-all ---
  console.log("\n=== Pattern 5: Catch-all ===");
  try {
    await client.documents.get("some-id");
  } catch (error) {
    if (error instanceof EdgeQuakeError) {
      // All SDK errors extend EdgeQuakeError
      console.log(
        `API error [${error.status}] ${error.code}: ${error.message}`,
      );
    } else {
      // Non-SDK error (e.g., TypeError from bad config)
      console.log("Unexpected error:", error);
    }
  }
}

main().catch(console.error);
